tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+		 * which the entire words of the filter bitmap are zero.
+		 */
+		if (filter)
+		{
+			if (filter[start_row / 64] == 0)
+			{
+				end_row = MIN(start_row + 64, n);
+				policy->stat_bulk_filtered_rows += 64;
+				continue;
+			}
+
+			for (end_row = start_row; end_row < n; end_row = MIN(end_row + 64, n))
+			{
+				if (filter[end_row / 64] == 0)
+				{
+					break;
+				}
+			}
+		}
+		else
+		{
+			end_row = n;
+		}
+		Assert(start_row <= end_row);
+		Assert(end_row <= n);
+
+		/*
+		 * Remember which aggregation states have already existed, and which we
+		 * have to initialize. State index zero is invalid, and state index one
+		 * is for null key. We have to initialize the null key state at the
+		 * first run.
+		 */
+		const uint32 last_initialized_state_index =
+			policy->table->members ? policy->table->members + 2 : 1;
+		uint32 next_unused_state_index = policy->table->members + 2;
+
+		/*
+		 * Match rows to aggregation states using a hash table.
+		 */
+		Assert((size_t) end_row <= policy->num_allocated_offsets);
+		switch ((int) key_column->decompression_type)
+		{
+			case DT_Scalar:
+				next_unused_state_index = fill_offsets_scalar(policy,
+															  batch_state,
+															  g->input_offset,
+															  next_unused_state_index,
+															  start_row,
+															  end_row,
+															  policy->offsets);
+				break;
+			case 8:
+				next_unused_state_index = fill_offsets_arrow_fixed_8(policy,
+																	 batch_state,
+																	 g->input_offset,
+																	 next_unused_state_index,
+																	 start_row,
+																	 end_row,
+																	 policy->offsets);
+				break;
+			case 4:
+				next_unused_state_index = fill_offsets_arrow_fixed_4(policy,
+																	 batch_state,
+																	 g->input_offset,
+																	 next_unused_state_index,
+																	 start_row,
+																	 end_row,
+																	 policy->offsets);
+				break;
+			case 2:
+				next_unused_state_index = fill_offsets_arrow_fixed_2(policy,
+																	 batch_state,
+																	 g->input_offset,
+																	 next_unused_state_index,
+																	 start_row,
+																	 end_row,
+																	 policy->offsets);
+				break;
+			default:
+				Assert(false);
+				break;
+		}
+
+		ListCell *aggdeflc;
+		ListCell *aggstatelc;
+
+		/*
+		 * Initialize the aggregate function states for the newly added keys.
+		 */
+		if (next_unused_state_index > last_initialized_state_index)
+		{
+			if (next_unused_state_index > policy->allocated_aggstate_rows)
+			{
+				policy->allocated_aggstate_rows = policy->allocated_aggstate_rows * 2 + 1;
+				forboth (aggdeflc, policy->agg_defs, aggstatelc, policy->per_agg_states)
+				{
+					VectorAggDef *agg_def = lfirst(aggdeflc);
+					lfirst(aggstatelc) =
+						repalloc(lfirst(aggstatelc),
+								 policy->allocated_aggstate_rows * agg_def->func.state_bytes);
```

Comment:
Wouldn't it be nicer if the agg_def->func had an allocation function instead? Then you'd just do `agg_def->func.alloc_states(num_states)` instead.

Reply:
I think it could be slightly confusing. The aggregate functions can't decide to allocate their memory in a different way, so this is not a part of their interface. No need to make a method that has the same single-line implementation for every aggregate.